File: app/src/main/java/algorithms/Sort.java
package algorithms;

import java.util.ArrayList;
import java.util.Collections;
import java.util.PriorityQueue;

public final class Sort {
  private Sort() {
    throw new UnsupportedOperationException("Utility class");
  }

  /**
   * Sorts a vector of integers in ascending order
   *
   * @param v The vector to be sorted
   * @throws NullPointerException if v is null
   */
  public static void sortVector(ArrayList<Integer> v) {
    if (v == null) {
      throw new NullPointerException("ArrayList cannot be null");
    }
    Collections.sort(v);
  }

  /**
   * Partitions a vector of integers around a pivot
   *
   * @param v           The vector to be partitioned
   * @param pivot_value The pivot value for partitioning
   * @throws NullPointerException if v is null
   */
  public static void dutchFlagPartition(ArrayList<Integer> v, int pivot_value) {
    if (v == null) {
      throw new NullPointerException("ArrayList cannot be null");
    }
    int next_value = 0;

    for (int i = 0; i < v.size(); i++) {
      if (v.get(i) < pivot_value) {
        Collections.swap(v, i, next_value);
        next_value++;
      }
    }

    for (int i = next_value; i < v.size(); i++) {
      if (v.get(i) == pivot_value) {
        Collections.swap(v, i, next_value);
        next_value++;
      }
    }
  }

  /**
   * Returns the largest n elements in a vector
   *
   * @param v The vector to be sorted
   * @param n The number of elements to return
   * @return A vector of the largest n elements in v
   * @throws NullPointerException if v is null
   * @throws IllegalArgumentException if n is negative or greater than the size of v
   */
  public static ArrayList<Integer> maxN(ArrayList<Integer> v, int n) {
    if (v == null) {
      throw new NullPointerException("ArrayList cannot be null");
    }
    if (n < 0) {
      throw new IllegalArgumentException("n cannot be negative: " + n);
    }
    if (n == 0 || n > v.size()) {
      throw new IllegalArgumentException("n must be between 1 and vector size (" + v.size() + "), got: " + n);
    }

    PriorityQueue<Integer> minHeap = new PriorityQueue<>();
    for (int i = 0; i < n; ++i) {
      minHeap.offer(v.get(i));
    }

    for (int i = n; i < v.size(); ++i) {
      if (v.get(i) > minHeap.peek()) {
        minHeap.poll();
        minHeap.offer(v.get(i));
      }
    }

    ArrayList<Integer> ret = new ArrayList<>(minHeap);
    Collections.sort(ret, Collections.reverseOrder()); // Sort in descending order
    return ret;
  }
}

File: src/sample_project/algorithms/sort.py
"""Sorting utilities and algorithms.

This module provides sorting utilities including basic sorting, the Dutch flag
partition algorithm for three-way partitioning, and heap-based selection of
the N largest elements from a list.
"""

import heapq


def sort_vector(data: list[int]) -> list[int]:
    """Sort a list of integers in ascending order.

    Creates and returns a new sorted list without modifying the original.
    Uses Python's built-in Timsort algorithm for efficient sorting.

    Args:
        data: The list of integers to be sorted.

    Returns:
        A new list containing all elements from data in ascending order.

    Raises:
        TypeError: If data is None.

    Note:
        Time complexity: O(n log n) where n is the length of the list.
        Space complexity: O(n) for the new sorted list.

    Example:
        >>> sort_vector([3, 1, 4, 1, 5, 9, 2, 6])
        [1, 1, 2, 3, 4, 5, 6, 9]
        >>> sort_vector([5, 2, 8, 1])
        [1, 2, 5, 8]
        >>> sort_vector([])
        []
    """
    if data is None:
        raise TypeError("List cannot be None")
    return sorted(data)


def dutch_flag_partition(arr: list[int], pivot: int) -> list[int]:
    """Partition a list around a pivot value using the Dutch flag algorithm.

    Implements the Dutch national flag problem solution, partitioning the
    list into three sections: elements less than pivot, elements equal to
    pivot, and elements greater than pivot. Returns a new partitioned list
    without modifying the original.

    The algorithm performs two passes:
    1. Move all elements < pivot to the front
    2. Move all elements == pivot after the < pivot section

    Args:
        arr: The list of integers to be partitioned.
        pivot: The pivot value for partitioning.

    Returns:
        A new list with elements partitioned into three sections:
        [elements < pivot][elements == pivot][elements > pivot]

    Raises:
        TypeError: If arr is None.

    Note:
        Time complexity: O(n) where n is the length of the list (two passes).
        Space complexity: O(n) for the new partitioned list.

    Example:
        >>> dutch_flag_partition([3, 5, 2, 6, 8, 1, 0, 5, 5], 5)
        [3, 2, 1, 0, 5, 5, 5, 6, 8]
        >>> dutch_flag_partition([1, 0, 2, 1, 0, 2, 1], 1)
        [0, 0, 1, 1, 1, 2, 2]
        >>> dutch_flag_partition([4, 4, 4], 4)
        [4, 4, 4]
    """
    if arr is None:
        raise TypeError("List cannot be None")

    # Create a copy to avoid modifying the original
    result = arr.copy()
    next_value = 0

    # First pass: move all elements < pivot to the front
    for i in range(len(result)):
        if result[i] < pivot:
            result[i], result[next_value] = result[next_value], result[i]
            next_value += 1

    # Second pass: move all elements == pivot after the < pivot section
    for i in range(next_value, len(result)):
        if result[i] == pivot:
            result[i], result[next_value] = result[next_value], result[i]
            next_value += 1

    return result


def max_n(data: list[int], n: int) -> list[int]:
    """Return the N largest elements from a list.

    Uses a min-heap of size N to efficiently find the largest N elements.
    The result is sorted in descending order (largest first).

    Args:
        data: The list of integers to search.
        n: The number of largest elements to return.

    Returns:
        A list of the N largest elements in descending order.

    Raises:
        TypeError: If data is None.
        ValueError: If n is negative, zero, or greater than the size of data.

    Note:
        Time complexity: O(n log k) where n is the list size and k is the
            number of elements to return.
        Space complexity: O(k) for the heap.

    Example:
        >>> max_n([3, 1, 4, 1, 5, 9, 2, 6], 3)
        [9, 6, 5]
        >>> max_n([10, 20, 30, 40, 50], 2)
        [50, 40]
        >>> max_n([7], 1)
        [7]
    """
    if data is None:
        raise TypeError("List cannot be None")
    if n < 0:
        raise ValueError(f"n cannot be negative: {n}")
    if n == 0 or n > len(data):
        raise ValueError(
            f"n must be between 1 and list size ({len(data)}), got: {n}"
        )

    # Build a min-heap with the first n elements
    min_heap = data[:n].copy()
    heapq.heapify(min_heap)

    # Process remaining elements
    for i in range(n, len(data)):
        if data[i] > min_heap[0]:
            heapq.heapreplace(min_heap, data[i])

    # Sort in descending order and return
    result = sorted(min_heap, reverse=True)
    return result
